codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -252,7 +252,7 @@ public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf dat
             Http2LocalFlowController flowController = flowController();
             try {
                 // If we should apply flow control, do so now.
-                if (shouldApplyFlowControl) {
+                if (shouldApplyFlowControl && bytesToReturn > 0) {
```

Comment:
> Which assertion was failing? In the new test?

I am guessing @adriancole was referring to this [assertion statement](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java#L312).  Please correct me if I'm wrong (or if there are other asserts that are also failing).